### PR TITLE
Use contract specific ABI encoding

### LINKED ISSUE #152: Replace encodeFunctionCall usages with methods.myMethod.encodeABI
Please, replace `eth.abi.encodeFunctionCall` usages in openst.js (`grep` the repo to find occurrences) with `methods.myMethod.encodeABI` calls (https://web3js.readthedocs.io/en/1.0/web3-eth-contract.html#methods-mymethod-encodeabi).

### PR DESCRIPTION
**Pull request by schemar**
I replaced all occurrences of `encodeFunctionCall` with the correct calls
to `methods.myMethod().encodeABI()`.

The existing integration tests assert that the behavior is as desired. I
did not see the value in creating unit tests that would basically mirror
the methods' code and "assert" that the same things are the same.

Fixes #152

### CODE REVIEW

**Comment by schemar** [comment]
I was thinking: I hope to achieve this without adding mosaic-contracts as a dependency. So that is a small thing I still want to change.

File: lib/helper/User.js [hunk 1]
```diff
@@ -146,23 +122,12 @@ class User {
...
-          },
-          {
-            type: 'bytes',
-            name: 'data'
-          }
-        ]
-      },
-      [this.proxyFactory, reducedModulesCreationData]
-    );
+    const createAndAddModules = new this.auxiliaryWeb3.eth.Contract(this.abiBinProvider.getABI('CreateAndAddModules'));
```
Comment (0xsarvesh) [inline lib/helper/User.js:126]: Regarding L100, instead of hardcoded json, should it be available in an ABI?

**Review by 0xsarvesh — COMMENTED** [review]
Changed looks good to me. 

Some feedback: 
1. `getCoGatewayRedeemExecutableData` doesn't have the unit as well as an integration test. 

2. Against the argument of not adding the unit tests, I think it will not duplicate the logic of the method. 

For instance, the unit test case of `getCoGatewayRedeemExecutableData` should assert a fixed value like `0x123dabcdef......` instead of reevaluating it. I think it's still valuable to add unit test cases. 
What do you think?

File: lib/helper/User.js (hunk 1 above)
Comment (schemar) [reply lib/helper/User.js:126]: Yes, it seems I overlooked it as I only grepped for `encodeFunctionCall`. But what *is* it? It uses a contract constructor. Where can I get the required ABI from? 😕

**Review by schemar — COMMENTED** [review]
Thanks @sarvesh-ost !
I need help with the hardcoded ABI input. Do you know where that is from/what that is?

Regarding you other comments:

> 1. `getCoGatewayRedeemExecutableData` doesn't have the unit as well as an integration test.
> 2. Against the argument of not adding the unit tests, I think it will not duplicate the logic of the method.
> 
> For instance, the unit test case of `getCoGatewayRedeemExecutableData` should assert a fixed value like `0x123dabcdef......` instead of reevaluating it. I think it's still valuable to add unit test cases.
> What do you think?

Hm 🤔 I would definitely like to test the method. But I don't know how to test it *properly.*

If the unit test case asserts a fixed value, then the test would break in case of an interface change even though the code is still correct. It would test the Web3js code and assert that Web3 does the ABI encoding correctly. For the current interface only.

On the other hand if the unit test case asserts a value that is generated in the unit test, then it would just be the same code. That obviously passes...

Another (weird) option would be to just call the method from the test and make sure it returns valid hex. It would fail if Web3js complains about the input values or structure and it would fail if it wouldn't produce any valid hex. Somehow, this feels weird although rationally it seems the most sensible of the options? 🤷‍♂️

So I am a little bit lost here 🤔 How would the method be tested correctly?

File: lib/helper/User.js (hunk 1 above)
Comment (0xsarvesh) [reply lib/helper/User.js:126]: Maybe some contract from safe contracts. Not sure though. 🤔

**Comment by 0xsarvesh** [comment]
> On the other hand if the unit test case asserts a value that is generated in the unit test, then it would just be the same code. That obviously passes...

Few points on the approach of asserting fixed value: 

1. In my opinion, it's a good thing if test complains about interface change. If the interface change is intentional, implementer needs to update test which is a conscious decision. 

2. It will not test the web3 per se, it also includes the logic of selecting correct abi and correct method. 
     

**Below selects correct ABI**
``` 
const delayedRecoveryModule = new this.auxiliaryWeb3.eth.Contract(
      this.abiBinProvider.getABI('DelayedRecoveryModule')
    );

```
**Below selects correct method** 
```
    const abiEncoded = delayedRecoveryModule.methods
      .setup(recoveryOwnerAddress, recoveryControllerAddress, recoveryBlockDelay)
      .encodeABI();
```

We should also get opinions from other team members on this.

**Comment by schemar** [comment]
> > On the other hand if the unit test case asserts a value that is generated in the unit test, then it would just be the same code. That obviously passes...
> 
> Few points on the approach of asserting fixed value:
> 
>     1. In my opinion, it's a good thing if test complains about interface change. If the interface change is intentional, implementer needs to update test which is a conscious decision.
> 
>     2. It will not test the web3 per se, it also includes the logic of selecting correct abi and correct method.
> 
> 
> **Below selects correct ABI**
> 
> ```
> const delayedRecoveryModule = new this.auxiliaryWeb3.eth.Contract(
>       this.abiBinProvider.getABI('DelayedRecoveryModule')
>     );
> ```
> 
> **Below selects correct method**
> 
> ```
>     const abiEncoded = delayedRecoveryModule.methods
>       .setup(recoveryOwnerAddress, recoveryControllerAddress, recoveryBlockDelay)
>       .encodeABI();
> ```
> 
> We should also get opinions from other team members on this.

OK I think this is the best option 👍 
@deepesh-kn @pgev any opinions?

**Comment by schemar** [comment]
I added the test, but I need feedback from @pgev to improve the mentioned comment from @sarvesh-ost.

File: lib/helper/User.js (hunk 1 above)
Comment (schemar) [reply lib/helper/User.js:126]: @pgev can you please help? Where can I get the ABI from without hard-coding it?

Comment (schemar) [reply lib/helper/User.js:126]: In line 100 that is ...

**Comment by schemar** [comment]
All addressed, ready for review again.

File: lib/helper/User.js (hunk 1 above)
Comment (schemar) [reply lib/helper/User.js:126]: According to @pgev the contract is not available and the ABI has to be hard-coded. I added a comment.

File: test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+describe('TokenHolder.getCoGatewayRedeemExecutableData()', () => {
+  let tokenHolder;
+
+  beforeEach(() => {
+    const web3 = new Web3();
+    const address = '0x0000000000000000000000000000000000000002';
+    tokenHolder = new TokenHolder(web3, address);
+  });
+
+  it('should construct with correct parameters', async () => {
```
Comment (jasonklein) [inline test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js]: I am confused by the caption for this test. But it looks like this is the caption for all of the positive TokenHolder unit tests. Is this just the practice that's been established here?

File: test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    const encodedAbi = tokenHolder.getCoGatewayRedeemExecutableData(
+      amount,
+      beneficiary,
+      gasPrice,
+      gasLimit,
+      nonce,
+      hashLock
+    );
+
+    const expectedEncodedAbi =
```
Comment (jasonklein) [inline test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js:36]: I think it would be helpful to add a brief comment here explaining the choice to go with a fixed value.

**Review by jasonklein — COMMENTED** [review]
2 small comments

File: test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js (hunk 2 above)
Comment (schemar) [reply test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js]: I *assumed.* @pgev ?

Comment (pgev) [reply test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js]: /cc @abhayks1 @gulshanvasnani

Comment (schemar) [reply test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js]: I can just fix it across the repository.

File: test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js (hunk 3 above)
Comment (schemar) [reply test/unit/TokenHolder/getCoGatewayRedeemExecutableData.js:36]: Will do.

**Review by 0xsarvesh — APPROVED** [review]
Nice 💯  

LGTM 👍
